<h2>1.2 Python Variables</h2><br><div style="text-align: justify">Variable assignment follows <b><code>variable_name = value</code></b>, where a single equal sign $=$ is an <b>assignment operator</b>. More on operators will be covered in the next section. Let's see a few examples of how we can do this.

In [2]:
# Let's create a variable called "a" and assign it the number 5
a = 5

Now if I use `a` in my Python script, Python will treat it as the number $5$.

In [3]:
# Adding variables
a + a

10

What happens on reassignment? Will Python let us write over it?

In [4]:
# Reassignment
a = 20

In [5]:
# Check
print(a)

20


<div style="text-align: justify">Yes! Python allows you to overwrite assigned variable names. We can also use the variables themselves when doing the reassignment.

<div style="text-align: justify">Since <b><code>a = 20</code></b> was the last assignment to our variable <b><code>a</code></b>, you can keep using <b><code>a</code></b> in place of the number <b><code>20</code></b>:

In [6]:
a = a + 5
print(a)

25


<div style="text-align: justify">Instead of writing <b><code>a+a</code></b>, Python has a built-in shortcut for these simple operations.<br><br>You can add, subtract, multiply and divide numbers with reassignment using <b><code>+=</code></b>, <b><code>-=</code></b>, <b><code>*=</code></b>, and <b><code>/=</code></b>, respectively.

In [7]:
a += 10

The above code will add **`10`** to the variable **`a`** every time you run that cell. 

Try it yourself, run it a few times and then run the below cell to see what's the value of **`a`**
.

In [8]:
print(a)

35


Below an example of a code that will double **`a`** every time that you run that cell.

In [9]:
a *= 2
print(a)

70


<h3>Determining variable type with <b><code>type()</code></b></h3><br><div style="text-align: justify">You can check what type of object is assigned to a variable using Python's built-in <b><code>type()</code></b> function. Common data types include:</div>


* **int** (for integer numbers)
* **float** (for floating point / all real numbers)
* **str** (for string/text)
* **bool** (for Boolean True/False)
* **list**
* **tuple**
* **dict** 
* **set**


<div class="alert alert-block alert-warning"><center><b>Always check the type of your variables when debugging!</b></center></div>

Below a few examples:

In [10]:
type(a)

int

In [11]:
float_var = 3.1415
type(float_var)

float

In [12]:
a = 0.3
b = 0.2
c = a - b
print(c)

0.09999999999999998


You probably noticed that Python wrote $0.09999999999999998$ instead of $0.1$ when calculating $0.3 - 0.2$. We will return to this later in this Notebook.

In [13]:
type(1 < 2)

bool

<div style="text-align: justify">Boolean variables can only take on two values: <b><code>True</code></b> or <b><code>False</code></b>. They are often used to check conditions.<br><br>More examples on this in Section 1.5.

In [14]:
1 < 2

True

In [15]:
# the variable from the first script

type(message)

str

<div style="text-align: justify"><b>Strings</b> are variables represented in between <b><code>' '</code></b> or <b><code>" "</code></b>.<br><br>They are a <b>sequence</b> of values, therefore you are able to access and manipulate every character individually.<br><br>This is done with the bracket operator <b><code>[]</code></b>, which works as an <b>index</b>.<br><br>Let's take a look at our first variable from this notebook: <b><code>message</code></b>.

In [16]:
message

'Hello world!'

In [17]:
message[1]

'e'

<b>What happened?</b>

<div style="text-align: justify">Why index <b><code>[1]</code></b> gave us the second letter? For most people, the first letter of the sentence <b><code>Hello world!</code></b> is <b><code>H</code></b> not <b><code>e</code></b>.</div>

So.. what happened?

In Python, indexing starts at **`[0]`**. **`H`** is the *zero-th* character of **`Hello world!`**.

In [18]:
message[0]

'H'

<div style="text-align: justify">You can also access the last value of a string using the index <b><code>[-1]</code></b>, the before-last using <b><code>[-2]</code></b> and so forth.. This will turn out to be very useful!

In [19]:
message[-1]

'!'

<div style="text-align: justify">Strings are <i>immutable</i>: you cannot reassign a new value for one of the characters. You will have to create a new string for that. Let's try to modify our string anyway:

In [20]:
message[0] = 'J'

TypeError: 'str' object does not support item assignment

This is an example of an error message. Python tells you that something is wrong with the code you just ran.
The final line gives the error itself. Above that is an indication of which line in the cell caused the error (helpful for cells with more complicated programs).

<div style="text-align: justify">You can add (i.e. concatenate) strings and characters. But it will create a new string, it will not modify the old one (since they are immutable).

In [21]:
message + message

'Hello world!Hello world!'

Now let's see how we can select a part of a string.

In [22]:
message[0] + message[1] + message[2] + message[3]

'Hell'


<div style="text-align: justify">   
A segment of a string is called a <b>slice</b>. Selecting a slice is similar to selecting a character. Using the operator <b><code>:</code></b> you select the first value that you want to <b>get</b>, and the first value you want to <b>leave out</b> of the slice, for example:<br><br></div>

<div style="text-align: justify">Let's say we want to write the word <b>Hell</b> using our variable <b><code>message</code></b>, without having to type as much as above.</div>

1) Which letter is the first we want to <b>get</b>? **`H`**, which has index **`[0]`**.

2) Which letter is the first we want to <b>leave out</b>? **`o`**, which has index **`[4]`**. So...

In [23]:
message[0:4]

'Hell'


<div class="alert alert-block alert-success"><center><b>Real life example: Analyzing satellite data<br><br></b>Sentinel data title is formatted as: <b>S1A_IW_SLC__1SDV_20181205T015821_20181205T015851_024884_02BD8C_8700</b><br>
    where each part means something, S1A means Sentinel-1A, IW means Interferometric Wide Swath<br> 20181205T015821 is a date, 2018 12 05, at 01h58m21s, etc.<br><br> Therefore, being able to manipulate this string is fundamental in order to organize and select satellite data. We'll come back to this in Exercise 1.5.4. </center></div>

***

<h3>Dynamic Typing</h3><br><div style="text-align: justify">
Python uses <b>dynamic typing</b>, meaning you can reassign variables to different data types. This makes Python very flexible in assigning data types; it differs from other languages that are <b>statically typed</b>, where each variable has a specified data type which cannot change.</div>

#### Pros and Cons of Dynamic Typing
##### Pros of Dynamic Typing
* very easy to work with
* faster development time

##### Cons of Dynamic Typing
* may result in unexpected bugs!
* you need to be aware of **`type()`**.

In [24]:
a = 5
print('Type of a is =',type(a))
a = 'string'
print('Type of a is =',type(a))

Type of a is = <class 'int'>
Type of a is = <class 'str'>


See, now **`a`** is no longer an **`int`** type but a **`str`** type

<h3>Casting types</h3><br><div style="text-align: justify">Sometimes you want to change the type of a variable. For example, there is no point in arithmetically adding a number to a string. These problems can sometimes be solved with casting. <b>Casting</b> is a procedure of changing variable type. Actually, you create a new variable with the requested data type using the variable you want to alter.</div><br>Examples are shown below.
    

In [25]:
string_number = '123'
print(string_number, type(string_number))

integer_number = int(string_number)
print(integer_number, type(integer_number))

123 <class 'str'>
123 <class 'int'>


<div style="text-align: justify">As you can see, both variables look the same in the output but their type now is different. Because of that, the cell below will result in an error.

In [26]:
string_number + 5

TypeError: can only concatenate str (not "int") to str

But the next cell will run normally.

In [27]:
integer_number + 5

128

<h3>Floating point numbers</h3>
<br>
We have seen that mathematics with integers (whole numbers) work as one might expect. But what about real numbers (with a decimal point)? 
On computers, these are called floating point numbers (where 'point' refers to the decimal point) or 'float' for short. 

An important thing to remember with floating point numbers is that doing mathematical operations with them is not exact, small rounding errors will appear. Even in simple calculations such as $0.3 - 0.1$. Try to run the cell below and see the result.</div>

In [28]:
0.3 - 0.1

0.19999999999999998

<div style="text-align: justify">It is not what you would expect to see, right? The result has an error of $\approx 10^{-15}$. In many cases these errors can be neglected, but be careful when comparing <b><code>float</code></b> and <b><code>int</code></b> numbers, as shown below. </div>

In [29]:
0.2 == 0.3 - 0.1

False

Indeed, $0.2 \neq 0.19999999999999998$.  A common advice is **Don't test floating point numbers for equality.** Testing for larger-than, smaller-than, is generally better.

<div style="text-align: justify">
    Within this course, in the end of each section you'll find exercises related to the subject that was just covered. There are three types of exercises: normal, fixing and searching.
    Normal exercises are straight forward exercises that you should be able to solve without much trouble.
    Fixing exercises are exercises where some piece of code is already written but contains an error. You need to debug it (find the cause of the error and fix it).
    Searching exercises are exercises that purposefully incorporate subjects that were not covered yet, in an attempt to encourage you to try and solve issues you haven't learned about yet. <br><br>
    Now let's practice!

```{important}
**Exercise 1.4.1**

You know that your taxes are the product of your income and the tax rate. Complete the line below to assign the correct calculation for the `my_taxes` variable. Use the `my_income` and `tax_rate` variables for that.

<div class="alert alert-block alert-info"><b>Exercise 1.2.1<br><br></b><div style="text-align: justify">You know that your taxes is the product of your income and the tax rate. Complete the line below to assign the correct calculation for the <code>my_taxes</code> variable. Use the <code>my_income</code> and <code>tax_rate</code> variables for that.<br> 
</div></div>

<div class="alert alert-block alert-warning"><b></b> The dotted lines (...) below mark the location where you should change the code yourself! </div>

In [30]:
my_income = 100
tax_rate = 0.1
my_taxes = ...
print(my_taxes)

Ellipsis


Use the cell block below to find out what is the **`type`** of the variable **`my_taxes`**

In [31]:
# use this line to check the type of the variable my_taxes
...



<div class="alert alert-block alert-info"><b>Exercise 1.2.2<br><br></b><div style="text-align: justify">What is the type of the result of the expression $3 + 1.5 + 4?$ <br>Assign the type to the variable beneath (e.g. if you think it is an integer then <code>expression_type = int</code>)
</div></div>

In [32]:
expression_type = ...


<div class="alert alert-block alert-info"><b>Exercise 1.2.3<br><br></b><div style="text-align: justify">Print the letter <code>'w'</code> of the sentence <code>'Hello world!'</code> using indexing of the variable <code>sentence</code>. <br>   
</div></div>

In [33]:
sentence = 'Hello world!'
letter_w = ...

print(letter_w)

Ellipsis


<div style="text-align: justify">
In the following exercise you will be using a piece of code, which was written by other people. It can contain concepts you haven't encountered yet but you will have to make it work! You need to practice with using code you don't understand completely, as you want to be able to use programs written by other people. </div><br>

<div style="text-align: justify">
In order to approach the problem, start from the point of error. If your error cannot be solved there, work your way backwards through the code. Try to understand the parts that are necessary to fix your problem, and avoid changing parts you know that work.</div>

<div class="alert alert-block alert-info">
<div style="text-align: justify">
    <b>(Fixing) Exercise 1.2.4<br><br>
    </b>There are 2 types of numbers in Python: integer numbers and floating point numbers. Both have overlapping and different applications. For example, counting rocks will require only integers (1 rock, 2 rocks, 3 rocks, etc), while their weight can be expressed in both <i>int</i> and <i>float</i> (depending on the required accuracy). <br>  
    Beneath you can see a small dataset of the measured daily temperatures and a filter function, which is used to remove incorrect measurements. Just run the cell, you don't have to change anything there and it is not important to understand it completely for now.
</div></div>


In [34]:
temp_data = [12.0, 13.5, 11.0, 12.7, 67.5, 11.8] # temperature measurements in Delft, in Celsius

# A function to remove erroneous measurements
def filter_errors(data, error_id):
    print('Data before filtering', data)
    
    del temp_data[error_id - 1] # deleting the erroneous measurement
    
    print('Data after filtering', data)

<div class="alert alert-block alert-info">
<div style="text-align: justify">
In the code cell below a student wanted to indicate the position of the erroneous measurement, so it can be later filtered out during data analysis. However, it wasn't removed. The filter is known to work flawlessly (no errors there), meaning that the mistake happened in the indication step. It is known that the corrupted sample is the 5th in the dataset (index 4 because the first element has index 0! Python has a zero-based indexing).
<br><br>
Your task is to fix it and remove the error!
    </div></div>

In [35]:
error_id = 4 # index of erroneous measurement

filter_errors(temp_data, error_id)

Data before filtering [12.0, 13.5, 11.0, 12.7, 67.5, 11.8]
Data after filtering [12.0, 13.5, 11.0, 67.5, 11.8]


<div class="alert alert-block alert-info">
<div style="text-align: justify">
<b>(Searching) Exercise 1.2.5 </b><br><br>
There are also other data types in Python, which were not mentioned in this part (in case they are not as useful or will be introduced later). For instance, <b>Complex</b> is one of them. The sad truth is that every software developer searches the web a lot. It's not because she/he is lazy or incompetent. It is because programming languages are constantly updated and some tools are limited to a narrow field of applications. It is impractical to learn everything at once, therefore every coder has learned how to look up the functionality their task requires. Therefore, it is crucial for you to learn how to search as well. Complex type is used in Python to represent complex numbers. If you haven't heard about them — a complex number is a number, which has a real and an imaginary part. For example, $x = 17 + 5i$. Here, $x$ is a complex number with a real part of $17$ and an imaginary part of $5$.<br><br> Your task is to create a variable <code>my_complex_num</code> of Complex type and assign a $3 + 2i$ value to it. For that you will have to Google a bit. Try to look for something like "Python complex variables". Python is very popular and you will be able to find everything you need. Make sure to filter the information you need — not everything you will find will be useful for this simple exercise.
</div></div>

In [36]:
my_complex_var = ...

print(my_complex_var)

Ellipsis


<div class="alert alert-block alert-danger"><b>Additional study material:</b>

* Official Python Documentation - https://docs.python.org/3/tutorial/introduction.html
* Think Python (2nd ed.) - Section 2

As an observation, web search quality has decreased over the last few years. For this reason, it may be helpful to search specific sites rather than using a general search engine. Some tips:
* The Think Python book, open it in a web browser or pdf reader, do text search (Control + F) (but it won't help you for this question...)
* The official Python documentation, or later when you start using *modules*, the documentation of the module in question.
* Use a general search engine, look at the Stack Overflow results
</div>